enhancements/cluster-logging/output-parameter-templates.md

```diff
@@ -0,0 +1,153 @@
+---
+title: output-parameter-templates
+
+authors:
+  - "@alanconway"
+
+reviewers:
+  - "@ewolinetz"
+  - "@lukas-vlcek"
+  - "@jcantril"
+  - "@sichvoge"
+  - "@vimalk78"
+
+approvers:
+  - "@jcantril"
+
+creation-date: 2021-04-21
+
+last-updated:
+
+status: provisional
+
+see-also:
+  - "/enhancements/cluster-logging/forwarding-json-structured-logs.md"
+---
+
+# Output Parameter Templates
+
+## Release Sign-off Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The ClusterLogForwarder.Outputs API can already set some output parameters from log record fields.
+
+For example, to set the Elasticsearch index for each log record to the value of kubernetes label "foo" on that record:
+
+``` yaml
+  elasticsearch:
+    indexKey: kubernetes.label.foo
+```
+
+This proposal allows multiple record fields to be _combined_ into an output value.
+
+For example, to set an index of the form "label_value+container_name":
+
+``` yaml
+  elasticsearch:
+    indexPattern: '{.kubernetes.label.foo}+{.kubernetes.container_name}'
+```
+
+The template pattern is [JSONpath](https://goessner.net/articles/JsonPath/).
+JSONpath is used by [kubernetes](https://kubernetes.io/docs/reference/kubectl/jsonpath)
+and [fluetnd](https://docs.fluentd.org/plugin-helper-overview/api-plugin-helper-record_accessor)
+and is relatively simple to learn.
+
+
+## Motivation
+
+Initial motivation comes from [forwarding JSON to Elasticsearch](forwarding-json-structured-logs.md).
+The popularity of JSON logging means that an application container be deployed in a pod with a sidecar logging an incompatible JSON log format.
+Adding `elasticsearch.indexPattern` allows JSON logs from different containers to be sent to separate indices.
+
+This feature can/will be used in any situation where per-log output parameters are derived from log record fields.
+For example `cloudwatch.groupPattern`, `loki.tenantPattern`, `kafka.topicPattern` and so on.
+
+### Goals
+
+Implement elasticsearch.indexPattern for the Elasticsearch output.
+
+### Non-Goals
+
+Other outputs will follow, but are not required to complete this proposal.
+
+## Proposal
+
+Add field `output.elasticsearch.indexPattern` of type string.
+Compute the index name for each log record by applying the `indexPattern` expression to the log record.
+
+### User Stories
+
+#### Send JSON logs from pods labeled app=myApp to separate indices based on container name
+
+``` yaml
+outputs:
+  - name: myApp
+    type: elasticsearch
+	indexPattern: '{.kubernetes.label.app.myApp}-{.kubernetes.container_name}'
```

Comment:
What is the behavior when  the pod does not have label: `app=myapp`?

Reply:
Needs clarifying. Will update.